Utilities: Port *most* utilities to Core::Stream

AtkinsSJ opened the pull request:
This is a first go to get the majority of simple (and less simple) cases done, leaving the ones that had me scratching my head for a while.

demostanis commented on Userland/Utilities/comm.cpp:
```diff
@@ -58,31 +61,38 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
-                return false;
-            }
-            file = file_or_error.value();
+    auto open_file = [](String const& path, auto& file, int file_number) {
```
Is this function really needed?

AtkinsSJ replied:
Without it, the logic would be duplicated. Line 88 calls this twice, once for each file.

demostanis replied:
Yep, but why are you making a function which handles errors manually instead of using `TRY(...)`? Also I'm unsure why you check if the path is a directory *after* opening it.

demostanis commented on Userland/Utilities/comm.cpp:
```diff
@@ -93,35 +103,41 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
-    String file1_line, file2_line;
+    String file1_line;
+    String file2_line;
+    auto buffer = TRY(ByteBuffer::create_uninitialized(4096));
```
Any reason you use 4096 instead of `PAGE_SIZE` like elsewhere?

AtkinsSJ replied:
Good point, that's better!

AtkinsSJ commented:
Replaced all but one use of `ByteBuffer::create_uninitialized()` with an `Array<>`. The remaining one is >32K so that felt a bit large for the stack. But maybe I'm worrying unnecessarily about that.

Also, using `PAGE_SIZE` as suggested.

demostanis commented on Userland/Utilities/cksum.cpp:
```diff
@@ -42,38 +43,48 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
+
     for (auto& path : paths) {
+        auto file_or_error = Core::Stream::File::open_file_or_standard_stream(path, Core::Stream::OpenMode::Read);
         auto filepath = (path == "-") ? "/dev/stdin" : path;
```
While checking if this line was actually needed, I found out that `cksum` is doing it wrong. You can't get the actual length of the standard input by `stat`ing it and reading st_size.
If you are motivated enough, you could fix this utility.

AtkinsSJ replied:
Nice catch! :^)

demostanis commented on Userland/Utilities/comm.cpp:
```diff
@@ -93,35 +103,41 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
-            if (print)
-                outln(fmt, line);
-        }
+        return strcmp(str1.characters(), str2.characters());
```
`str1 == str2`?

AtkinsSJ replied:
`strcmp` returns a number that tells you whether str1 goes before or after str2 alphabetically, which is used on lines 140, 145 and 151.

demostanis commented on Userland/Utilities/comm.cpp:
```diff
@@ -93,35 +103,41 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
+        }
+        if (read_file2) {
+            auto can_read_file2_line = file2->can_read_line();
+            if (can_read_file2_line.is_error() || !can_read_file2_line.value())
```
`return !can_read_file?_line.is_error() && can_read_file?_line.value()` looks better IMO.

AtkinsSJ replied:
We only want to return false if it's wrong - if it's not, we carry on to the next part. (Otherwise, if `read_file1` and `read_file2` were both true, we'd only check file 1.)

demostanis commented on Userland/Utilities/markdown-check.cpp:
```diff
@@ -233,14 +234,22 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
         auto file = file_or_error.release_value();
-        auto content_buffer = file->read_all();
+
+        auto content_buffer_or_error = file->read_all();
```
doesn't `TRY(file->read_all())` work here?

demostanis replied:
But you're returning if there's an error. Since when does that carry on to the next file? 🤨

AtkinsSJ replied:
Ohhhh you're right. :man_facepalming:

I think I wrote it like this to match the original error handling, which prints an error saying which file had the problem. That's useful when it's dealing with several files. Hopefully we'll figure out how to do that with ErrorOr eventually.

Sorry if I've seemed a bit dismissive of your feedback - my intention here was to quickly port things to Core::Stream without doing any other refactoring or fixes. It's true most of these could be modified to be better, but then this PR would grow even more chonky. :sweat_smile: It's already big enough to hang around quite a while.

AtkinsSJ commented:
Since last time:
- Fixed the merge request
- Fixed `cksum` trying to `stat()` stdin

Apparently Buggie still doesn't like `w` as a category name. :thonk:

ADKaster commented:
I merged a separate `checksum` PR that has conflicts w/your changes :)

AtkinsSJ commented:
> I merged a separate `checksum` PR that has conflicts w/your changes :)

Thanks! They already ported it to Core::Stream so I've dropped my commit that did that. No other changes, apart from rebasing.

davidot commented on Userland/Utilities/comm.cpp:
```diff
@@ -58,31 +61,38 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
+            return false;
+        }
+
+        if (Core::File::is_directory(path)) {
+            warnln("Failed to open file{} '{}': is a directory", file_number, path);
+            return false;
+        }
```
This might slightly change the behavior if you have a directory named `-`.
Since that gets STDIN but then here we do check if it's a directory where we only did that check for `path != "-"` before

davidot commented on Userland/Utilities/gron.cpp:
```diff
@@ -41,16 +41,11 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
-        file = Core::File::standard_input();
-    else
-        file = TRY(Core::File::open(path, Core::OpenMode::ReadOnly));
+    auto file = TRY(Core::Stream::File::open(path, Core::Stream::OpenMode::Read));
```
This no longer read from STDIN if no path is null I think

Although using `open_standard...` technically also changes behavior as `-` would be treated as STDIN

davidot reviewed (commented):
Tried to look at only thing that are stream related.
Found two things

davidot reviewed (approved):
I didn't go over it in detail as I did that last time so changes LGTM.
I'm sure we'll find any problems when people actually use the programs.

AtkinsSJ commented:
Changes: Fixed a conflict, which was just that Andreas removed an `outln()` from `lsirq`, so this should still be good to go!
